src/ORM/DataQuery.php

```diff
@@ -723,6 +721,56 @@ public function conjunctiveGroup()
         return new DataQuery_SubGroup($this, 'AND', $clause);
     }
 
+    /**
+     * Adds a Common Table Expression (CTE), aka WITH clause.
+     *
+     * Use of this method should usually be within a conditional check against DB::get_conn()->supportsCteQueries().
+     *
+     * @param string $name The name of the WITH clause, which can be referenced in any queries UNIONed to the $query
+     * and in this query directly, as though it were a table name.
+     * @param string[] $cteFields Aliases for any columns selected in $query which can be referenced in any queries
+     * UNIONed to the $query and in this query directly, as though they were columns in a real table.
+     * NOTE: If $query is a DataQuery, then cteFields must be the names of real columns on that DataQuery's data class.
+     * @param string|string[] $onClause The "ON" clause (escaped SQL statement) for an INNER JOIN on the query.
```

Comment:
Seems like we should just not have the $onClause param at, seems like we're trying to be helpful by making it a tiny bit easier to add an INNER JOIN at the cost of impurifying this method?

Honestly I think if you're using this API at this point you've probably already written you complex query with a CTE manually using DB::query() and now you're just trying to do it "properly" with more ORM

Reply:
> Seems like we should just not have the $onClause param at, seems like we're trying to be helpful by making it a tiny bit easier to add an INNER JOIN at the cost of impurifying this method?

Inner join is (from what I've seen while I was looking into how to implement this) the most common type of join used in CTE queries. So this caters for the most common scenario while _allowing_ alternatives.

> Honestly I think if you're using this API at this point you've probably already written you complex query with a CTE manually using DB::query() and now you're just trying to do it "properly" with more ORM

I don't know where that assumption comes from. It hasn't been supported until MySQL 8. There's a lot of room to use this API in core code now that we're introducing it, and we _don't_ have any purely raw SQL queries using CTEs right now.

Reply:
Sorry what I meant is that because CTE's are complex, personally I'd write it out as raw SQL first using DB::query() and then convert it to use the ORM later and validate it against my raw query. I wouldn't start out with alterDataQuery() :-)

I still don't think we should be polluting this method with this inner join "helper", after all you can just go $dataQuery->with(<withParams>)->innerJoin(<innerJoinParams>) to do the same thing

Reply:
Inner join will be by far the most common join for a CTE query. Part of the point of having higher level abstractions is that they take care of common scenarios for you, which is what is happening here.

To put it into perspective:
- The only time you'd want a left join is if you want _all_ of the records in the main query, and are only _adding_ to the result set with the CTE. I can't think of a single scenario where you'd want to to that.
- You'd use a right join only when your CTE is filling in some gaps in your `DataQuery` result set. There's an example of that in the unit tests, where the CTE gets the full date range between the min and max dates in the result set, and adds a bunch of _null_ results for the missing dates. There's really only a handful of cases where you'd want to do something like this, and in probably 90% of those cases you'd more likely just do it directly in PHP instead.
- Inner join would be used when you want to effectively filter your result set based on the CTE query. This covers scenarios like getting the ancestors/descendents for a record, or getting all archived versions of records (where finding "wasDeleted" versions in the latest_versions set is the non-recursive CTE, which gets inner joined on ID = version ID).

In the scenarios I've identified so far where we could use CTEs to simplify or optimise queries in our codebase, _none_ of them would use a left or right join.

Reply:
Even if we keep the $onClause param at very least we kind of need to get rid of array support because it's simply inconsistent with the rest of the ORM, e.g. [DataList::innerJoin()](https://github.com/silverstripe/silverstripe-framework/blob/5/src/ORM/DataList.php#L775) which should be more "helpful" given it's the API projects actually use only supports the string param for $onClause not an array params

Once the with() method is down to only supporting string params then at that point all we're doing is ` $this->query->addInnerJoin($name, $onClause);` so at that point you may as well just go `->with()->innerJoin($onClause)`

Reply:
I really don't understand where you're coming from with this, and especially I don't understand why you keep putting quotes around "helpful" lol. I can't see a world where this isn't useful additional abstraction.
However, given that you also seemingly can't see where I'm coming from, I'll remove this piece of code. Just don't be surprised if in a year's time we have a lot of this
```php
$joinField = DataObject::getSchema()->sqlColumnForField($dataQuery->dataClass(), 'JoinField');
$cteJoinField = Convert::symbol2sql([$cteName, 'CTEField']);
$dataQuery->with(...)->innerJoin("$joinField = $cteJoinField");
```
that we want to abstract away.

Reply:
Also 
> e.g. [DataList::innerJoin()](https://github.com/silverstripe/silverstripe-framework/blob/5/src/ORM/DataList.php#L775) which should be more "helpful" given it's the API projects actually use only supports the string param for $onClause not an array param

I absolutely agree that the `DataList` join abstractions are shit and should be made more useful. If adding this array syntax support to those would help you accept this I'd very gladly raise a PR to do that.

Reply:
I have removed the `$onClause` parameter, though I still say this makes the abstraction considerably less useful.